### ISSUE TITLE
[BUG] 2.15 using qat_deflate with default docker image crashes node because of missing library

### ISSUE THREAD

**Issue by sandervandegeijn** [opensearch-project/custom-codecs#168]
### Describe the bug

Trying to use the qat_deflate compression. According to the docs it should be there from 2.14 on. This is not correct btw, in 2.14.0 it can't be used, gives an instant error. Created a PR for the docs for that one. 

In 2.15 is does not throw the error of not supporting the codec but it does crash the node.

### Related component

Storage

### To Reproduce

```
PUT /telegraf-rabbitmq-dip-tst-2024.07.24-qat
{
  "settings": {
    "index": {
      "codec": "qat_deflate"
    }
  }
}
```

Then reindex:

```
POST /_reindex
{
  "source": {
    "index": "telegraf-rabbitmq-dip-tst-2024.07.24"
  },
  "dest": {
    "index": "telegraf-rabbitmq-dip-tst-2024.07.24-qat"
  }
}
```

### Expected behavior

Do not crash

### Additional Details

Base 2.15.0 docker image with the s3 plugin installed.

Log:

```
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot [2024-07-25T20:02:25,225][ERROR][o.o.b.OpenSearchUncaughtExceptionHandler] [opensearch-data-nodes-hot-1] fatal error in thread [opensearch[opensearch-data-nodes-hot-1][write][T#3]], exiting
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot java.lang.UnsatisfiedLinkError: /tmp/opensearch-9601680627269700390/libqat-java12656953293004161880.so: libqatzip.so.3: cannot open shared object file: No such file or directory
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at java.base/jdk.internal.loader.NativeLibraries.load(Native Method) ~[?:?]
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at java.base/jdk.internal.loader.NativeLibraries$NativeLibraryImpl.open(NativeLibraries.java:331) ~[?:?]
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at java.base/jdk.internal.loader.NativeLibraries.loadLibrary(NativeLibraries.java:197) ~[?:?]
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at java.base/jdk.internal.loader.NativeLibraries.loadLibrary(NativeLibraries.java:139) ~[?:?]
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at java.base/java.lang.ClassLoader.loadLibrary(ClassLoader.java:2418) ~[?:?]
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at java.base/java.lang.Runtime.load0(Runtime.java:852) ~[?:?]
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at java.base/java.lang.System.load(System.java:2025) ~[?:?]
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at com.intel.qat.Native.lambda$loadLibrary$1(Native.java:99) ~[?:?]
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at java.base/java.security.AccessController.doPrivileged(AccessController.java:319) ~[?:?]
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at com.intel.qat.Native.loadLibrary(Native.java:102) ~[?:?]
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at com.intel.qat.InternalJNI.<clinit>(InternalJNI.java:17) ~[?:?]
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at com.intel.qat.QatZipper.<clinit>(QatZipper.java:97) ~[?:?]
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at org.opensearch.index.codec.customcodecs.QatZipperFactory.createInstance(QatZipperFactory.java:34) ~[?:?]
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at org.opensearch.index.codec.customcodecs.QatZipperFactory.createInstance(QatZipperFactory.java:166) ~[?:?]
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at org.opensearch.index.codec.customcodecs.QatCompressionMode$QatCompressor.<init>(QatCompressionMode.java:96) ~[?:?]
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at org.opensearch.index.codec.customcodecs.QatCompressionMode.newCompressor(QatCompressionMode.java:75) ~[?:?]
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at org.apache.lucene.codecs.lucene90.compressing.Lucene90CompressingStoredFieldsWriter.<init>(Lucene90CompressingStoredFieldsWriter.java:118) ~[lucene-core-9.10.0.jar:9.10.0 695c0ac84508438302cd346a812cfa2fdc5a10df - 2024-02-14 16:48:06]
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at org.apache.lucene.codecs.lucene90.compressing.Lucene90CompressingStoredFieldsFormat.fieldsWriter(Lucene90CompressingStoredFieldsFormat.java:140) ~[lucene-core-9.10.0.jar:9.10.0 695c0ac84508438302cd346a812cfa2fdc5a10df - 2024-02-14 16:48:06]
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at org.opensearch.index.codec.customcodecs.Lucene99QatStoredFieldsFormat.fieldsWriter(Lucene99QatStoredFieldsFormat.java:124) ~[?:?]
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at org.apache.lucene.index.StoredFieldsConsumer.initStoredFieldsWriter(StoredFieldsConsumer.java:50) ~[lucene-core-9.10.0.jar:9.10.0 695c0ac84508438302cd346a812cfa2fdc5a10df - 2024-02-14 16:48:06]
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at org.apache.lucene.index.StoredFieldsConsumer.startDocument(StoredFieldsConsumer.java:57) ~[lucene-core-9.10.0.jar:9.10.0 695c0ac84508438302cd346a812cfa2fdc5a10df - 2024-02-14 16:48:06]
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at org.apache.lucene.index.IndexingChain.startStoredFields(IndexingChain.java:535) ~[lucene-core-9.10.0.jar:9.10.0 695c0ac84508438302cd346a812cfa2fdc5a10df - 2024-02-14 16:48:06]
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at org.apache.lucene.index.IndexingChain.processDocument(IndexingChain.java:566) ~[lucene-core-9.10.0.jar:9.10.0 695c0ac84508438302cd346a812cfa2fdc5a10df - 2024-02-14 16:48:06]
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at org.apache.lucene.index.DocumentsWriterPerThread.updateDocuments(DocumentsWriterPerThread.java:263) ~[lucene-core-9.10.0.jar:9.10.0 695c0ac84508438302cd346a812cfa2fdc5a10df - 2024-02-14 16:48:06]
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at org.apache.lucene.index.DocumentsWriter.updateDocuments(DocumentsWriter.java:425) ~[lucene-core-9.10.0.jar:9.10.0 695c0ac84508438302cd346a812cfa2fdc5a10df - 2024-02-14 16:48:06]
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at org.apache.lucene.index.IndexWriter.updateDocuments(IndexWriter.java:1558) ~[lucene-core-9.10.0.jar:9.10.0 695c0ac84508438302cd346a812cfa2fdc5a10df - 2024-02-14 16:48:06]
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at org.apache.lucene.index.IndexWriter.updateDocument(IndexWriter.java:1843) ~[lucene-core-9.10.0.jar:9.10.0 695c0ac84508438302cd346a812cfa2fdc5a10df - 2024-02-14 16:48:06]
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at org.apache.lucene.index.IndexWriter.addDocument(IndexWriter.java:1483) ~[lucene-core-9.10.0.jar:9.10.0 695c0ac84508438302cd346a812cfa2fdc5a10df - 2024-02-14 16:48:06]
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at org.opensearch.index.engine.InternalEngine.addDocs(InternalEngine.java:1281) ~[opensearch-2.15.0.jar:2.15.0]
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at org.opensearch.index.engine.InternalEngine.indexIntoLucene(InternalEngine.java:1217) ~[opensearch-2.15.0.jar:2.15.0]
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at org.opensearch.index.engine.InternalEngine.index(InternalEngine.java:1011) ~[opensearch-2.15.0.jar:2.15.0]
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at org.opensearch.index.shard.IndexShard.index(IndexShard.java:1215) ~[opensearch-2.15.0.jar:2.15.0]
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at org.opensearch.index.shard.IndexShard.applyIndexOperation(IndexShard.java:1160) ~[opensearch-2.15.0.jar:2.15.0]
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at org.opensearch.index.shard.IndexShard.applyIndexOperationOnPrimary(IndexShard.java:1051) ~[opensearch-2.15.0.jar:2.15.0]
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at org.opensearch.action.bulk.TransportShardBulkAction.executeBulkItemRequest(TransportShardBulkAction.java:625) ~[opensearch-2.15.0.jar:2.15.0]
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at org.opensearch.action.bulk.TransportShardBulkAction$2.doRun(TransportShardBulkAction.java:471) ~[opensearch-2.15.0.jar:2.15.0]
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at org.opensearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:941) ~[opensearch-2.15.0.jar:2.15.0]
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at org.opensearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:52) ~[opensearch-2.15.0.jar:2.15.0]
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144) ~[?:?]
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642) ~[?:?]
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at java.base/java.lang.Thread.run(Thread.java:1583) [?:?]
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot fatal error in thread [opensearch[opensearch-data-nodes-hot-1][write][T#3]], exiting
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot java.lang.UnsatisfiedLinkError: /tmp/opensearch-9601680627269700390/libqat-java12656953293004161880.so: libqatzip.so.3: cannot open shared object file: No such file or directory
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at java.base/jdk.internal.loader.NativeLibraries.load(Native Method)
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at java.base/jdk.internal.loader.NativeLibraries$NativeLibraryImpl.open(NativeLibraries.java:331)
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at java.base/jdk.internal.loader.NativeLibraries.loadLibrary(NativeLibraries.java:197)
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at java.base/jdk.internal.loader.NativeLibraries.loadLibrary(NativeLibraries.java:139)
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at java.base/java.lang.ClassLoader.loadLibrary(ClassLoader.java:2418)
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at java.base/java.lang.Runtime.load0(Runtime.java:852)
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at java.base/java.lang.System.load(System.java:2025)
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at com.intel.qat.Native.lambda$loadLibrary$1(Native.java:99)
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at java.base/java.security.AccessController.doPrivileged(AccessController.java:319)
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at com.intel.qat.Native.loadLibrary(Native.java:102)
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at com.intel.qat.InternalJNI.<clinit>(InternalJNI.java:17)
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at com.intel.qat.QatZipper.<clinit>(QatZipper.java:97)
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at org.opensearch.index.codec.customcodecs.QatZipperFactory.createInstance(QatZipperFactory.java:34)
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at org.opensearch.index.codec.customcodecs.QatZipperFactory.createInstance(QatZipperFactory.java:166)
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at org.opensearch.index.codec.customcodecs.QatCompressionMode$QatCompressor.<init>(QatCompressionMode.java:96)
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at org.opensearch.index.codec.customcodecs.QatCompressionMode.newCompressor(QatCompressionMode.java:75)
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at org.apache.lucene.codecs.lucene90.compressing.Lucene90CompressingStoredFieldsWriter.<init>(Lucene90CompressingStoredFieldsWriter.java:118)
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at org.apache.lucene.codecs.lucene90.compressing.Lucene90CompressingStoredFieldsFormat.fieldsWriter(Lucene90CompressingStoredFieldsFormat.java:140)
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at org.opensearch.index.codec.customcodecs.Lucene99QatStoredFieldsFormat.fieldsWriter(Lucene99QatStoredFieldsFormat.java:124)
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at org.apache.lucene.index.StoredFieldsConsumer.initStoredFieldsWriter(StoredFieldsConsumer.java:50)
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at org.apache.lucene.index.StoredFieldsConsumer.startDocument(StoredFieldsConsumer.java:57)
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at org.apache.lucene.index.IndexingChain.startStoredFields(IndexingChain.java:535)
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at org.apache.lucene.index.IndexingChain.processDocument(IndexingChain.java:566)
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at org.apache.lucene.index.DocumentsWriterPerThread.updateDocuments(DocumentsWriterPerThread.java:263)
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at org.apache.lucene.index.DocumentsWriter.updateDocuments(DocumentsWriter.java:425)
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at org.apache.lucene.index.IndexWriter.updateDocuments(IndexWriter.java:1558)
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at org.apache.lucene.index.IndexWriter.updateDocument(IndexWriter.java:1843)
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at org.apache.lucene.index.IndexWriter.addDocument(IndexWriter.java:1483)
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at org.opensearch.index.engine.InternalEngine.addDocs(InternalEngine.java:1281)
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at org.opensearch.index.engine.InternalEngine.indexIntoLucene(InternalEngine.java:1217)
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at org.opensearch.index.engine.InternalEngine.index(InternalEngine.java:1011)
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at org.opensearch.index.shard.IndexShard.index(IndexShard.java:1215)
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at org.opensearch.index.shard.IndexShard.applyIndexOperation(IndexShard.java:1160)
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at org.opensearch.index.shard.IndexShard.applyIndexOperationOnPrimary(IndexShard.java:1051)
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at org.opensearch.action.bulk.TransportShardBulkAction.executeBulkItemRequest(TransportShardBulkAction.java:625)
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at org.opensearch.action.bulk.TransportShardBulkAction$2.doRun(TransportShardBulkAction.java:471)
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at org.opensearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:941)
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at org.opensearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:52)
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
opensearch-data-nodes-hot-1 opensearch-data-nodes-hot 	at java.base/java.lang.Thread.run(Thread.java:1583)

```

**Comment by sandervandegeijn** [comment]
> > > > > opensearch-cluster-1  | Caused by: java.lang.UnsupportedOperationException: Unsupported OS/arch, cannot find /com/intel/qat/linux/aarch64/libqat-java.so. Please try building from source.
> > > 
> > > 
> > > That seems like a different issue: `opensearch-cluster-1 | Caused by: java.lang.UnsupportedOperationException: Unsupported OS/arch, cannot find /com/intel/qat/linux/aarch64/libqat-java.so. Please try building from source`
> > 
> > 
> > Confirmed this one on a Mac with Apple Silicon running the image under docker desktop as well. Didn't expect that error either, from the docs: it should fall back to the software implementation instead of the hardware accelerated one, but it should still work.
> > Seperate bug?
> 
> Yeah that probably is a separate issue because in custom-codecs we see this
> 
> ```
> [opensearch@3b3758158c66 test]$ unzip qat-java-1.1.1.jar
> Archive:  qat-java-1.1.1.jar
>    creating: META-INF/
>   inflating: META-INF/MANIFEST.MF
>    creating: com/
>    creating: com/intel/
>    creating: com/intel/qat/
>    creating: com/intel/qat/linux/
>    creating: com/intel/qat/linux/amd64/
>    creating: META-INF/maven/
>    creating: META-INF/maven/com.intel.qat/
>    creating: META-INF/maven/com.intel.qat/qat-java/
>   inflating: com/intel/qat/Native.class
>   inflating: com/intel/qat/QatZipper$Mode.class
>   inflating: com/intel/qat/QatException.class
>   inflating: com/intel/qat/QatZipper$QatCleaner.class
>   inflating: com/intel/qat/QatDecompressorInputStream.class
>   inflating: com/intel/qat/package-info.class
>   inflating: com/intel/qat/QatCompressorOutputStream.class
>   inflating: com/intel/qat/InternalJNI.class
>   inflating: com/intel/qat/QatZipper.class
>   inflating: com/intel/qat/QatZipper$PollingMode.class
>   inflating: com/intel/qat/QatZipper$Algorithm.class
>   inflating: com/intel/qat/linux/amd64/libqat-java.so
>   inflating: META-INF/maven/com.intel.qat/qat-java/pom.xml
>   inflating: META-INF/maven/com.intel.qat/qat-java/pom.properties
>   inflating: module-info.class
> ```
> 
> It seems like it does not support arm64 at this point.

Still it should fall back to a pure software implementation right? Should I open a seperate issue in the custom-codecs repo?

**Comment by andrross** [comment]
These look like similar issues...the QAT codecs do not gracefully handle the case where it cannot be loaded (either due to incompatible hardware or missing library).

I would propose the following options for the 2.16 release:

1. Fix the issue so that the service gracefully fails with a proper error message at index creation time if the codec is not supported. (probably not feasible to do this right in the short time that we have, but happy to be proven wrong)
2. Remove the QAT codecs (this is a breaking change with no work-around available for any existing users that are using one of these codecs)
3. Add a setting to opt-in to use these new codecs, but make them unavailable by default (this is a breaking change but gives a work-around to re-enable them. any user is still at risk of a node crash if they enable this setting in the wrong environment)

**Comment by getsaurabh02** [comment]
@andrross  thanks for sharing the options. 
Do we know if any existing users can successfully upgrade their clusters with indices on the QAT codec without encountering failures, and will it work seamlessly?
If yes, then can we move forward with a one or combination of #2 and #3 above where enablement for new indices can be prevented?

**Comment by dblock** [comment]
@andrross Why isn't there a catch all in `IndexWriter`, maybe `initStoredFieldsWriter` on failure to initialize one of these codecs?

**Comment by peterzhuamazon** [comment]
Able to reproduce:
```

fatal error in thread [opensearch[039c85055a8f][write][T#4]], exiting
java.lang.UnsatisfiedLinkError: /tmp/opensearch-18442301159217191627/libqat-java8257943536944747477.so: libqatzip.so.3: cannot open shared object file: No such file or directory
```

**Comment by sarthakaggarwal97** [comment]
> Do we know if any existing users can successfully upgrade their clusters with indices on the QAT codec without encountering failures, and will it work seamlessly?

If the qat codec is unavailable, I doubt the shards will be green. The users can change the codec before upgrading (force merge to 1 segment, so that all the segment has a old / stable codec), and then upgrade.

> Add a setting to opt-in to use these new codecs, but make them unavailable by default 

I think this is what we wanted always, but we couldn't come up with the consensus on how to mark these codecs experimental. Discussions over here https://github.com/opensearch-project/OpenSearch/pull/13992

I'm good with 3rd option, if we have can come up with a mitigation plan to fix the codecs. If we do not see that happening soon, I will vote for 2nd option.

**Comment by andrross** [comment]
> @andrross Why isn't there a catch all in `IndexWriter`, maybe `initStoredFieldsWriter` on failure to initialize any of these codecs?

@dblock These are `java.lang.Error` instances, not exceptions. It is generally unsafe to catch Errors as it usually indicates that the JVM is not able to continue operating properly. I suspect the right solution here is to introspect at runtime whether these codecs are available, otherwise don't register them.

**Comment by sarthakaggarwal97** [comment]
adding @mulugetam to the discussion (contributor for QAT codec)

**Comment by peterzhuamazon** [comment]
Hi @sarthakaggarwal97 @mulugetam ,

Do we know if user needs to explicitly install `libqatzip.so.3` on their machine?
In our docker image we also do not have this lib it seems, and I cant seem to find it from any existing repositories.

If so, regardless of which option we need to add it to the docker image.
Please let me know the installation of these or you would include it with custom-codecs plugin.

Thanks.

**Comment by dblock** [comment]
> > @andrross Why isn't there a catch all in `IndexWriter`, maybe `initStoredFieldsWriter` on failure to initialize any of these codecs?
> 
> @dblock These are `java.lang.Error` instances, not exceptions. It is generally unsafe to catch Errors as it usually indicates that the JVM is not able to continue operating properly. I suspect the right solution here is to introspect at runtime whether these codecs are available, otherwise don't register them.

But this is `java.lang.UnsatisfiedLinkError`. An explicit `loadLibrary` for codecs along with a `catch` for this should be ok, no?

**Comment by backslasht** [comment]
>     1. Fix the issue so that the service gracefully fails with a proper error message at index creation time if the codec is not supported. (probably not feasible to do this right in the short time that we have, but happy to be proven wrong)
I agree, we shouldn't return the codecs if they are not supported in that platform.

>     3. Add a setting to opt-in to use these new codecs, but make them unavailable by default (this is a breaking change but gives a work-around to re-enable them. any user is still at risk of a node crash if they enable this setting in the wrong environment)
Not a requirement for 2.16. But, we should implement this sooner, today the installation of custom-codes gets all the codecs in it causing issues like this.

**Comment by andrross** [comment]
> But this is java.lang.UnsatisfiedLinkError. An explicit loadLibrary for codecs along with a catch for this should be ok, no?

@dblock Yes, @sarthakaggarwal97 has implemented something like this. But I don't think we should have a general catch-all in the core layer.

**Comment by andrross** [comment]
One thing to note here is that I believe from reading the [EC2 documentation](https://aws.amazon.com/ec2/instance-types/c7i/) that only the `metal` sizes of the m7i, r7i, and c7i instance types have the QAT hardware acceleration. We don't use those instance sizes in any of our testing infrastructure (to my knowledge) so I don't think we're actually testing this codec anywhere in practice. Some of the documentation suggests that this should fallback to software acceleration but in my test the `isQatAvailable` always seems to evaluate to false, even on a stock Ubuntu EC2 instance using an intel processor.

**Comment by andrross** [comment]
Fixed by #169. Closing.

**Comment by sandervandegeijn** [comment]
Thanks guys
